internal/etl/transform/extend_values.go

```diff
@@ -2,24 +2,52 @@ package transform
 
 import (
 	"encoding/csv"
+	"fmt"
 	"os"
+	"slices"
 	"strings"
 )
 
-func generateFieldValuesSubstitutions(fieldValuesByName map[string]any, substitutionFile string) ([]map[string]any, error) {
-	substitutions, err := readSubstitutionsFile(substitutionFile)
+// Return slice of fieldValuesByName
+func generateAllFieldValues(fieldValuesByName map[string]any, substitutionsFile, synonymsFile string) ([]map[string]any, error) {
+	substitutions, err := readSubstitutionsFile(substitutionsFile)
 	if err != nil {
 		return nil, err
 	}
-	var fieldValuesByNameExtensions = make(map[string][]string)
+	synonyms, err := readSubstitutionsFile(synonymsFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var fieldValuesByNameWithAllValues = make(map[string][]string)
 	for key, value := range fieldValuesByName {
-		valueSubstitutions, err := applySubstitutions(value.(string), substitutions)
+		valueLower := strings.ToLower(value.(string))
+
+		// Get all substitutions
+		substitutedValues, err := extentValues([]string{valueLower}, substitutions)
+		if err != nil {
+			return nil, err
+		}
+		// Get all synonyms for these substituted values
+		// one way
+		synonymsValuesOneWay, err := extentValues(substitutedValues, synonyms)
+		if err != nil {
+			return nil, err
+		}
+		// reverse way
+		reverseSynonyms, err := reverseMap(synonyms)
 		if err != nil {
 			return nil, err
 		}
-		fieldValuesByNameExtensions[key] = valueSubstitutions
+		allValues, err := extentValues(synonymsValuesOneWay, reverseSynonyms)
+		if err != nil {
+			return nil, err
+		}
+
+		// Create map with for each key a slice of []values
+		fieldValuesByNameWithAllValues[key] = allValues
 	}
-	return generateAllFieldValuesByName(fieldValuesByNameExtensions), err
+	return generateAllFieldValuesByName(fieldValuesByNameWithAllValues), err
 }
 
 // Transform a map[string][]string into a []map[string]string using the cartesian product, i.e.
```

Comment:
Zit niet in deze change maar een recursive functie kan je bijna altijd ook zonder recursie schrijven. `generateCombinations` is recursief, heb een poging gedaan deze zonder recursie op te lossen:

```
func generateCombinations(keys []string, values [][]string) []map[string]any {
	if len(keys) == 0 || len(values) == 0 {
		return nil
	}
	result := []map[string]any{{}} // contains empty map so the first iteration works
	for keyDepth := 0; keyDepth < len(keys); keyDepth++ {
		var newResult []map[string]any
		for _, entry := range result {
			for _, val := range values[keyDepth] {
				newEntry := make(map[string]any)
				for k, v := range entry {
					newEntry[k] = v
				}
				newEntry[keys[keyDepth]] = val
				newResult = append(newResult, newEntry)
			}
		}
		result = newResult
	}
	return result
}
```

En de test:

```
func Test_generateCombinations(t *testing.T) {
	type args struct {
		keys   []string
		values [][]string
	}
	tests := []struct {
		name string
		args args
		want []map[string]any
	}{
		{"Single key, single value", args{[]string{"key1"}, [][]string{{"value1"}}}, []map[string]any{{"key1": "value1"}}},
		{"Single key, slice of values", args{[]string{"key1"}, [][]string{{"value1", "value2"}}}, []map[string]any{{"key1": "value1"}, {"key1": "value2"}}},
		{"Two keys, two single values", args{[]string{"key1", "key2"}, [][]string{{"value1"}, {"value2"}}}, []map[string]any{{"key1": "value1", "key2": "value2"}}},
		{"Two keys, slice + single value", args{[]string{"key1", "key2"}, [][]string{{"value1", "value2"}, {"value3"}}}, []map[string]any{{"key1": "value1", "key2": "value3"}, {"key1": "value2", "key2": "value3"}}},
		{"Two keys, two slices values", args{[]string{"key1", "key2"}, [][]string{{"value1", "value2"}, {"value3", "value4"}}}, []map[string]any{{"key1": "value1", "key2": "value3"}, {"key1": "value1", "key2": "value4"}, {"key1": "value2", "key2": "value3"}, {"key1": "value2", "key2": "value4"}}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assert.Equalf(t, tt.want, generateCombinations(tt.args.keys, tt.args.values), "generateCombinations(%v, %v)", tt.args.keys, tt.args.values)
		})
	}
}
```

Reply:
Mooi en clean! Heb hem overgenomen